pkg/awsutils/awsutils.go

```diff
@@ -343,21 +329,85 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata() error {
 	}
 	log.Debugf("Found vpc-ipv4-cidr-block: %s ", cache.vpcIPv4CIDR)
 
-	// retrieve vpc-ipv4-cidr-blocks
+	// spawing go routine and retrieving vpc-ipv4-cidr-block
+	waitCh := make(chan string)
+	errorCh := make(chan error)
+	go wait.Forever(func() {
+		cache.initSGIDs(mac, waitCh, errorCh)
+		cache.initVPCIPv4CIDRs(mac, waitCh, errorCh)
+	}, 30*time.Second)
+
+	select {
+	case <-waitCh:
+		return nil
+	case errs := <-errorCh:
+		return errs
+	}
+}
+
+// initSGIDs retrieves security groups
+func (cache *EC2InstanceMetadataCache) initSGIDs(mac string, waitCh chan string, errorCh chan error) {
+	metadataSGIDs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataSGs)
+	if err != nil {
+		awsAPIErrInc("GetMetadata", err)
+		log.Errorf("Failed to retrieve security-group-ids data from instance metadata service")
+		errorCh <- errors.Wrap(err, "get instance metadata: failed to retrieve security-group-ids")
+		return
+	}
+
+	sgIDs := strings.Fields(metadataSGIDs)
+	for _, sgID := range sgIDs {
+		log.Debugf("Found security-group id: %s", sgID)
+		cache.appendIfMissing(sgID)
+	}
+	waitCh <- "Security Groups retrieved"
+}
+
+// initVPCIPv4CIDRs retrieves VPC IPv4 CIDR blocks
+func (cache *EC2InstanceMetadataCache) initVPCIPv4CIDRs(mac string, waitCh chan string, errorCh chan error) {
 	metadataVPCIPv4CIDRs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataVPCcidrs)
+
 	if err != nil {
 		awsAPIErrInc("GetMetadata", err)
 		log.Errorf("Failed to retrieve vpc-ipv4-cidr-blocks from instance metadata service")
-		return errors.Wrap(err, "get instance metadata: failed to retrieve vpc-ipv4-cidr-block data")
+		errorCh <- errors.Wrap(err, "get instance metadata: failed to retrieve vpc-ipv4-cidr-blocks data")
+		return
 	}
 
 	vpcIPv4CIDRs := strings.Fields(metadataVPCIPv4CIDRs)
-
 	for _, vpcCIDR := range vpcIPv4CIDRs {
 		log.Debugf("Found VPC CIDR: %s", vpcCIDR)
-		cache.vpcIPv4CIDRs = append(cache.vpcIPv4CIDRs, aws.String(vpcCIDR))
+		cache.appendIfMissing(vpcCIDR)
+	}
+	waitCh <- "VPC CIDR blocks retrieved"
+}
+
+func (cache *EC2InstanceMetadataCache) appendIfMissing(input string) {
+	if strings.Contains(input, "sg-") {
```

Comment:
I'm not sure this function is needed. Is it not easier to just set `cache.securityGroups` to whatever the metadata returns? Also, this is the eth0 (primary MAC) only. It is used in `createENI` when setting the SG to use. If we modify and read this in different threads, we need to lock it. Something similar to the `UnmanagedENISet`. Maybe we should rename that type to `StringSet` and use it for both?